src/AdoNetCore.AseClient/AseCommandBuilder.cs

```diff
@@ -0,0 +1,574 @@
+                                catch (FormatException)
+                                {
+                                    // Swallow the FormatException if it's because the value is "NULL"
+                                    if (!string.Equals(reader.GetString(scaleOrdinal).Trim(), "NULL", StringComparison.OrdinalIgnoreCase))
+                                    {
+                                        throw;
+                                    }
+                                }
+                            }
+                        }
+
+                        var size = 0;
+                        if (!reader.IsDBNull(lengthOrdinal))
+                        {
+                            size = reader.GetInt16(lengthOrdinal);
+                        }
+                            
+
+                        var parameter = new AseParameter(parameterName, dbType, size) { IsNullable = isNullable, Precision = precision, Scale = scale };
+
+                        if (isFirstParameter)
+                        {
+                            parameter.Direction = ParameterDirection.ReturnValue;
+                            isFirstParameter = false;
+                        }
+                        else
+                        {
+                            try
+                            {
+                                // Get the mode.
+                                if (reader.FieldCount > 21 && !reader.IsDBNull(21))
+                                {
+                                    if (string.Equals(reader.GetString(21), "out", StringComparison.OrdinalIgnoreCase))
+                                    {
+                                        parameter.Direction = ParameterDirection.Output;
+                                    }
+                                }
+                            }
+                            catch (IndexOutOfRangeException) // TODO - this shouldn't happen with the FieldCount check above.
+                            {
+                            }
+                        }
+                        command.Parameters.Add(parameter);
+                    }
+                }
+            }
+        }
+
+        private static bool HasPrecisionAndScale(AseDbType aseDbType)
+        {
+            switch (aseDbType)
+            {
+                case AseDbType.Double:
+                case AseDbType.Decimal:
+                case AseDbType.Money:
+                case AseDbType.Numeric:
+                    return true;
+                default:
+                    return false;
+            }
+        }
+
+        protected override void ApplyParameterInfo(DbParameter parameter, DataRow row, StatementType statementType, bool whereClause)
+        {
+            var aseParameter = (AseParameter) parameter;
+            aseParameter.AseDbType = GetAseDbTypeFromTdsName(row["DataTypeName"].ToString());
+            //aseParameter.AseDbType = (AseDbType) row[SchemaTableColumn.ProviderType];
```

Comment:
From `AseDataReader`:
```
row[providerType] = (DbType)column.DataType;
```
`column.DataType` is an `AseDbType` so I reckon you'd be able to cast, rather than parse the string

Reply:
There's something weird going on with it. I get the value 175 back - which doesn't correspond to anything on DbType. The DbType enum is from 0-27. The DataTypeName column gives me "LongVarChar" which I can process... thoughts?

Reply:
Ah whoops, that corresponds to `TdsDataType.TDS_LONGCHAR` (0xAF).
Looks like I got it wrong in the `AseDataReader`, it's casting from `TdsDataType` to `DbType` when it should probably be doing a conversion (via a not-yet-existent TypeMap method)...

Reply:
Fixed now